doc/llvm_obfuscation_tutorial.rst

```diff
@@ -0,0 +1,979 @@
+        Value *RhsTot = Builder.CreateMul(RhsSquare, prime2);
+        registerInteger(*RhsTot);
+
+        // comp
+        Value *comp =
+            Builder.CreateICmp(CmpInst::Predicate::ICMP_EQ, LhsTot, RhsTot);
+        registerInteger(*comp);
+        Value *castComp = Builder.CreateZExt(comp, ReplacedType);
+        registerInteger(*castComp);
+
+        return castComp;
+      }
+    };
+    }
+
+DOOOOOOOOOOOOOOOOOOOOONE!
+
+Let's try this awesome pass, if we use it on the last version of our test code we get:
+
+.. code:: llvm
+
+    ; Function Attrs: nounwind uwtable
+    define i32 @main() #0 {
+      %1 = alloca i32, align 4
+      %a = alloca i32, align 4
+      store i32 0, i32* %1
+      store i32 2, i32* %a, align 4
+      %2 = call i32 @puts(i8* getelementptr inbounds ([13 x i8]* @.str, i32 0, i32 0))
+      %3 = load i32* %a, align 4
+      %4 = mul nsw i32 %3, 3
+      store i32 %4, i32* %a, align 4
+      %5 = and i32 %3, 7
+      %6 = or i32 %5, 2
+      %7 = mul i32 %6, %6
+      %8 = mul i32 %7, 719
+      %9 = and i32 %2, 7
+      %10 = or i32 %9, 8
+      %11 = mul i32 %10, %10
+      %12 = mul i32 %11, 397
+      %13 = icmp eq i32 %8, %12
+      %14 = zext i1 %13 to i32
+      ret i32 %14
+    }
+
+Look at the assignments %5 to %14, looks familiar? We have successfully obfuscated the ``return 0`` instruction with the expression we gave at the beginning.
+
+But there are a few important things left to read, so stay tunned!
+
+
+You didn't think it would be that easy?
++++++++++++++++++++++++++++++++++++++++
+
+The optimizer is your enemy
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+So far we have not tried to optimize our code.
+But the compiler can optimize away some of your obfuscations and turn the code back to its original form.
+Our obfuscation depends on some rather complex arithmetic properties so we are safe but you should keep in mind that the compiler might be working against you.
+
+Even though our arithmetic is optimization-proof the rest of the code is not. The optimizer can still modify your code and delete all the candidate variables for x and y. If you want to see this effect, comment out the ``puts`` call in our test code and add the -O3 flag to your compilation command.
+
+You should get this:
+
+.. code:: llvm
+
+    ; Function Attrs: nounwind readnone uwtable
+    define i32 @main() #0 {
+      ret i32 0
+    }
+
+In this case the compiler has optimized out ``a`` which was the only integer available for the obfuscation.
+This explains why the obfuscation aborted.
+
+Even if it is frustrating it is not a real problem, since the compiler won't delete all the potential integer in a real code.
+However this is very annoying when writing tests.
+The easiest work-around is to declare ``volatile`` the variables you don't want to be optimized out.
+
+You might think that not using the optimizer is a good solution but:
+    * If your obfuscation can't resist an optimizer, it won't resist reverse engineers.
```

Comment:
Yeah this is explained earlier, it's more like general remark.
If you think I should clarify let me know.